src/main/java/algorithms/curated170/medium/SparseMatrixMultiplicationWithHashMap.java

```diff
@@ -0,0 +1,37 @@
+package algorithms.curated170.medium;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class SparseMatrixMultiplicationWithHashMap {
+    public static int[][] solution(int[][] A, int[][] B) {
+        int m = A.length;
+        int n = A[0].length;
+        int l = B[0].length;
+        int[][] C = new int[m][l];
+        Map<Integer, Map<Integer, Integer>> map = new HashMap<>();
+
+        for(int i=0; i< n; i++){
+            map.put(i, new HashMap<Integer, Integer>());
+            for(int j = 0; j<l; j++){
+                if(B[i][j] != 0){
+                    map.get(i).put(j, B[i][j]); //We store all numbers that are not zero.
+                }
+            }
+        }
```

Comment:
Buradaki sıkıştırma kodunu compress diye ayrı bir methoda taşımak ve bunu hem A'ya hem B'ye uygulamak, ardından Linear Algebra sınıfındaki method ile çarpmak daha modüler bir yapı sağlayabilir.

Reply:
İkinci döngüde ` if(A[i][j] != 0)` diye bir kontrol olduğu için böyle bir şeye gerek kalacağını sanmıyorum. Her türlü A matriksini de döngüleyeceğimiz için onu da kendi döngüsünde haritalayıp o sırada 0 olanları elemek işlem farkı vermemeli.

Reply:
Hız bakımından bir fark sağlamaz. Sâdece daha modüler olurdu. Ama karâr senin. 
Bana kalırsa kod bu hâli ile de yeterli.

Reply:
Eğer modülerlik için düşünürsek o iki döngüyü mü yeni yönteme almalıyız?

Reply:
Eğer sıkıştırma mantığını yanlış anlamadı isem, bir Matrix bir HashMap'in içine şu şekilde sıkıştırılabilir:


```
     private HashMap<Integer, Map<Integer, Integer>> compress(int [][] mat) {
         HashMap<Integer, Map<Integer, Integer>> compressedMatrix = new HashMap<>();
        for (int i = 0; i < mat.length; i++) {
            for (int j = 0; j< mat[0].length; j++) {
                if (mat[i][j] != 0) {
                 
                       compressedMatrix.get(i).put(j,mat[i][j]);  
                }
            }
        }
        return compressedMatrix;
    }
```

Buna iki matrix'i de gönderip sıkıştırdıktan sonra, bu matrixleri HashMap üzerinden çarpabilirdik. Tabiî, Linear Algebra sınıfını da HashMap alacak şekilde düzenlememiz veyâ onu kullanmadan direkt HashMap üzerinden çarpmamız gerekirdi.

Bana kalırsa çok mühim bir refactör işlemi değil. Ama modülerliğin daha yalın ve anlaşılır oluşuna bir örnek olarak aklımızda bulunabilir.

Reply:
B matriksi için yaptığımız sıkıştırmayı bu yöntemle ekledim. Hashmap'lerin matriks çarpımını almayı da deniyorum.

Reply:
Tamamdır, çok güzel oldu.
Ben de bu hâli ile LeetCode'da deneyeyim.

Reply:
Kodumuz bu hâli ile çok iyi çalışıyor.
Yalnız şu sarıladığım yeri unutmuşum:

![image](https://user-images.githubusercontent.com/60903744/115956976-7ff4c080-a508-11eb-9e93-fc8216c6f688.png)

Reply:
Yalnız HashMap'leri çarpan çözümü ekledim. Burada mapA.get(i).get(j) yapmak yerine array'den değer almak daha mı iyi olur?

Reply:
Şu ân hem zamân hem mekan bakımından çok kaliteli oldu.

![image](https://user-images.githubusercontent.com/60903744/115957083-2d67d400-a509-11eb-84fa-b50427b2419e.png)


Muhtemelen HashMap'ten almak da gayet hızlıdır. Key ile hashlendiği için, array ile yarışır bu konuda.

Reply:
![image](https://user-images.githubusercontent.com/63192680/115957194-b717a180-a509-11eb-8900-b8cc7652b5ae.png)
Göstermiyor ama şu anki çözüm önceki çözümden de hızlı. 
Ayrıca buradaki `System.out.println(java.time.Clock.systemUTC().instant()); işlemi milisaniyelere kadar zamanı gösterdiği için yerel işlemlerde iyi bir performans karşılaştırma yöntemi olabilir.

Reply:
Süper görünüyor.